File: r152/src/piece152.py
#!/usr/bin/env python3
"""Round 152 -- an INDEPENDENT Python checker for the marked piece capacities.

The twin of r152/src/piece152.c.  It reads the same L6-PIECECERT-1 bytes and
re-establishes both directions itself:

    lower   the stored witness replays as a legal piece with that mask
    upper   an exhaustive search for cap+1 ports under the mask finds nothing
    (-1)    the whole space for that mask is exhausted with nothing found

It imports the catalogue from r152/src/checker152.py, which builds it from
string algebra and reads nothing from the production solvers, and it reads no
production table: the only values it prunes with are the ones it has already
certified earlier in the same run.

usage: piece152.py --cert <cert.txt> [--report r.json] [--node-cap N]
                   [--compare outputs/rr_l6_marked_capacity_table_144.json]
"""
from __future__ import annotations
import argparse, hashlib, json, sys, time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
from checker152 import HEX, ORB, PHASE, FREE, PAID, N                 # noqa: E402

HEXCAP = 120


class Piece:
    def __init__(self, certified, node_cap):
        self.cert = dict(certified)        # {(b,d): C(b,d,0,0)} already proved
        self.cap = node_cap
        self.nodes = 0
        self.found = None
        self.trail = None

    def ub(self, tok, d):
        best = HEXCAP
        for (kb, kd), c in self.cert.items():
            if kb >= tok and kd >= d and c < best:
                best = c
        return best

    @staticmethod
    def feas(phm, corb, tok, dmax):
        us = sorted((5 - len(ph) for q, ph in phm.items() if q != corb),
                    reverse=True)
        return sum(us[tok:]) <= dmax

    def search(self, b, dmax, fp, lp, target):
        self.nodes = 0
        self.found = None
        self.trail = None
        phm = {ORB[0]: {PHASE[0]}}
        hexu = {HEX[0]}
        path = [0]

        def rec(cur, corb, ports, tok, firstlen, curlen, nblk, deficit):
            self.nodes += 1
            if self.cap and self.nodes > self.cap:
                raise TimeoutError
            if deficit <= dmax and ports >= target:
                fl = firstlen if nblk > 1 else curlen
                if (not fp or fl < 5) and (not lp or curlen < 5):
                    self.found, self.trail = ports, list(path)
                    raise StopIteration
            if not self.feas(phm, corb, tok, dmax):
                return
            reach = min(ports + self.ub(tok, dmax - deficit + 4 + 5 * tok) - 1,
                        ports + (HEXCAP - len(hexu)))
            if reach < target:
                return
            t = FREE[cur]
            if HEX[t] not in hexu and PHASE[t] not in phm[corb]:
                phm[corb].add(PHASE[t])
                hexu.add(HEX[t])
                path.append(t)
                rec(t, corb, ports + 1, tok, firstlen, curlen + 1, nblk,
                    deficit - 1)
                path.pop()
                hexu.discard(HEX[t])
                phm[corb].discard(PHASE[t])
            for t in PAID[cur]:
                if HEX[t] in hexu:
                    continue
                q = ORB[t]
                ph = phm.get(q)
                fresh = ph is None
                if not fresh:
                    if PHASE[t] in ph:
                        continue
                    if tok < 1:
                        continue
                if fresh:
                    phm[q] = {PHASE[t]}
                else:
                    ph.add(PHASE[t])
                hexu.add(HEX[t])
                path.append(t)
                rec(t, q, ports + 1, tok - (0 if fresh else 1),
                    firstlen if nblk > 1 else curlen, 1, nblk + 1,
                    deficit + (4 if fresh else -1))
                path.pop()
                hexu.discard(HEX[t])
                if fresh:
                    del phm[q]
                else:
                    ph.discard(PHASE[t])

        sys.setrecursionlimit(10000)
        try:
            rec(0, ORB[0], 1, b, 0, 1, 1, 4)
        except StopIteration:
            return "FOUND"
        except TimeoutError:
            return "CAP"
        return "EXHAUSTED"

    @staticmethod
    def replay(b, dmax, fp, lp, cap, w):
        if len(w) != cap:
            return False, "witness length != cap"
        if not w or w[0] != 0:
            return False, "the piece must start at 123456"
        phm = {ORB[0]: {PHASE[0]}}
        hexu = {HEX[0]}
        norb, deficit, tok, corb = 1, 4, 0, ORB[0]
        firstlen, curlen, nblk = 0, 1, 1
        for i in range(len(w) - 1):
            u, t = w[i], w[i + 1]
            if not 0 <= t < N:
                return False, "port out of range"
            if HEX[t] in hexu:
                return False, f"step {i} revisits a hexagon"
            q = ORB[t]
            if PHASE[t] in phm.get(q, ()):
                return False, f"step {i} reuses a phase"
            if t == FREE[u]:
                if q != corb:
                    return False, f"step {i}: clean E left its orbit"
                phm[q].add(PHASE[t])
                deficit -= 1
                curlen += 1
            elif t in PAID[u]:
                fresh = q not in phm
                if not fresh:
                    tok += 1
                phm.setdefault(q, set()).add(PHASE[t])
                if fresh:
                    norb += 1
                    deficit += 4
                else:
                    deficit -= 1
                firstlen = firstlen if nblk > 1 else curlen
                curlen, nblk, corb = 1, nblk + 1, q
            else:
                return False, f"step {i} is neither a clean E nor a paid joint"
            hexu.add(HEX[t])
        if tok > b:
            return False, f"token budget exceeded: {tok} > {b}"
        if deficit > dmax:
            return False, f"deficit budget exceeded: {deficit} > {dmax}"
        if deficit != 5 * norb - len(w):
            return False, "deficit identity violated"
        fl = firstlen if nblk > 1 else curlen
        if fp and fl >= 5:
            return False, "the mask requires a partial FIRST block"
        if lp and curlen >= 5:
            return False, "the mask requires a partial LAST block"
        return True, dict(ports=len(w), orbits=norb, deficit=deficit, tok=tok,
                          first_block=fl, last_block=curlen)


def read_pcert(text):
    toks = [t for line in text.splitlines()
            for t in line.split("#")[0].split()
            if not t.startswith("L6-PIECECERT")]
    out, i = [], 0
    while i < len(toks):
        if toks[i] != "pcell":
            raise SystemExit(f"expected 'pcell' at token {i}, saw {toks[i]!r}")
        b, d, fp, lp, cap, n = (int(x) for x in toks[i + 1:i + 7])
        w = [int(x) for x in toks[i + 7:i + 7 + n]]
        if len(w) != n:
            raise SystemExit("witness truncated")
        out.append(dict(b=b, d=d, fp=fp, lp=lp, cap=cap, witness=w))
        i += 7 + n
    return out


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--cert", required=True)
    ap.add_argument("--report", default=None)
    ap.add_argument("--node-cap", type=int, default=2_000_000_000)
    ap.add_argument("--compare", default=None)
    a = ap.parse_args()
    raw = Path(a.cert).read_bytes()
    if not raw.startswith(b"L6-PIECECERT-1"):
        raise SystemExit("not an L6-PIECECERT-1 file")
    entries = read_pcert(raw.decode())
    compare = None
    if a.compare:
        t = json.loads(Path(a.compare).read_text())["tables"]
        compare = {}
        for b, tab in t.items():
            for k, v in tab.items():
                d, mask = k.split("|")
                compare[(int(b), int(d), int(mask[0]), int(mask[1]))] = v

    certified, rows, ok = {}, [], True
    t0 = time.time()
    for e in entries:
        b, d, fp, lp, cap, w = (e["b"], e["d"], e["fp"], e["lp"], e["cap"],
                                e["witness"])
        key = f"{b}|{d}|{fp}{lp}"
        p = Piece(certified, a.node_cap)
        t1 = time.time()
        row = dict(cell=key, b=b, d=d, fp=fp, lp=lp, cap=cap)
        if cap < 0:
            r = p.search(b, d, fp, lp, 1)
            row["nodes"] = p.nodes
            row["status"] = {"FOUND": "DISAGREE", "CAP": "UNKNOWN_CAP",
                             "EXHAUSTED": "EXACT_CERTIFIED"}[r]
        else:
            good, info = (Piece.replay(b, d, fp, lp, cap, w) if w
                          else (True, dict(ports=cap)))
            if not good:
                row.update(status="WITNESS_BAD", detail=info)
            else:
                r = p.search(b, d, fp, lp, cap + 1)
                row["nodes"] = p.nodes
                row["status"] = {"FOUND": "DISAGREE", "CAP": "UNKNOWN_CAP",
                                 "EXHAUSTED": ("EXACT_CERTIFIED" if w
                                               else "UPPER_CERTIFIED")}[r]
        if row["status"] in ("EXACT_CERTIFIED", "UPPER_CERTIFIED"):
            if fp == 0 and lp == 0 and cap >= 0:
                certified[(b, d)] = cap
        else:
            ok = False
        if compare is not None:
            c = compare.get((b, d, fp, lp))
            row["claimed_elsewhere"] = c
            if c is not None and c != cap:
                row["status"] = "MISMATCH_VS_CLAIM"
                ok = False
        row["seconds"] = round(time.time() - t1, 2)
        rows.append(row)
        print(f"  {key:>12} cap={cap:<4} {row['status']:<18} "
              f"nodes={row.get('nodes', 0):<12,} {row['seconds']}s", flush=True)
    out = dict(cert=a.cert, cert_sha256=hashlib.sha256(raw).hexdigest(),
               node_cap=a.node_cap, cells=len(rows),
               certified=sum(r["status"] in ("EXACT_CERTIFIED",
                                             "UPPER_CERTIFIED") for r in rows),
               total_nodes=sum(r.get("nodes", 0) for r in rows),
               seconds=round(time.time() - t0, 1), all_certified=ok, rows=rows)
    if a.report:
        Path(a.report).parent.mkdir(parents=True, exist_ok=True)
        Path(a.report).write_text(json.dumps(out, indent=1) + "\n")
    print(json.dumps({k: v for k, v in out.items() if k != "rows"}, indent=1))
    return 0 if ok else 1


if __name__ == "__main__":
    sys.exit(main())

File: r152/src/piece152.c
/* Round 152 -- producer and checker for the MARKED PIECE capacities C(b,d,mask).
 *
 * The chain capacities are not the only capacity table the L = 870/871 census
 * leans on.  Dropping the piece model leaves 52 rows at L = 871 and one at
 * L = 870 unclosed (r152/certs/piece_needed_152.json), so the piece table is
 * load-bearing too and needs the same treatment.
 *
 * DEFINITION.  A PIECE for (b, d) is a walk v_0 = 123456, v_1, ... over the
 * 720 words in which
 *   * every step is a clean E (= tau) or one of the five paid joints,
 *   * every word lands on a hexagon not used before (the piece is hex-simple),
 *   * no phase of an orbit is used twice,
 *   * a step that re-enters an already opened orbit costs one token, and a
 *     clean E never costs one; at most b tokens in all,
 *   * the deficit 5*(orbits opened) - (ports) is at most d.
 * A BLOCK is a maximal run of clean E steps.  With fp, lp in {0,1},
 *
 *     C(b, d, fp, lp) = max ports over such pieces whose FIRST block is
 *                       partial (< 5 ports) if fp = 1, and whose LAST block is
 *                       partial if lp = 1; a mask bit 0 constrains nothing.
 *                       -1 when no piece meets the constraint.
 *
 * Certificate format L6-PIECECERT-1, plain text:
 *     pcell <b> <d> <fp> <lp> <cap> <nports>
 *     <nports port indices>
 * cap = -1 carries no witness and claims the mask is unreachable, which the
 * checker establishes by exhausting the whole space for that mask.
 *
 * build: cc -O2 -o piece152.exe piece152.c
 * run:   ./piece152.exe check   <cert.txt> [node_cap]
 *        ./piece152.exe produce <claims.txt> <out.txt> [node_cap]
 *        claims.txt: one "<b> <d> <fp> <lp> <cap>" per line, in certification
 *        order.
 */
#define CHECKER152_NO_MAIN
#include "checker152.c"

/* ---------------------------------------------------------------- the bound */
#define PT 8
#define PD 64
static int PUB[PT][PD];
static void pub_init(void) {
    for (int t = 0; t < PT; ++t) for (int d = 0; d < PD; ++d) PUB[t][d] = 120;
}
static void pub_add(int b, int d, int val) {     /* a certified (b,d,0,0) */
    for (int t = 0; t <= b && t < PT; ++t)
        for (int dd = 0; dd <= d && dd < PD; ++dd)
            if (val < PUB[t][dd]) PUB[t][dd] = val;
}
static int pub(int tok, int d) {
    if (tok < 0) tok = 0;
    if (d < 0) d = 0;
    if (tok >= PT || d >= PD) {
        fprintf(stderr, "pub index out of range: %d %d\n", tok, d);
        exit(3);                                            /* fail closed */
    }
    return PUB[tok][d];
}

/* --------------------------------------------------------------- the search */
static int PB, PDMAX, PFP, PLP, PTARGET;
static int PTRAIL[900];
static int PWIT[900], PWITN;

static void prec(int cur, int corb, int ports, int tok, int firstlen,
                 int curlen, int nblk) {
    ++NODES;
    if (NODECAP && NODES > NODECAP) { CAPPED = 1; return; }
    if (deficit <= PDMAX && ports >= PTARGET) {
        int fl = (nblk > 1) ? firstlen : curlen;
        if ((!PFP || fl < 5) && (!PLP || curlen < 5)) {
            FOUND = 1; PWITN = ports;
            for (int i = 0; i < ports; ++i) PWIT[i] = PTRAIL[i];
            return;
        }
    }
    if (!feas(tok, corb)) return;                 /* DMAX is read by feas */
    int reach = ports + pub(tok, PDMAX - deficit + 4 + 5 * tok) - 1;
    int reach2 = ports + (NH - hexcount);          /* a piece is hex-simple */
    if (reach2 < reach) reach = reach2;
    if (reach < PTARGET) return;

    int t = FREE_T[cur];
    if (!hexu[HEXID[t]] && !(phm[corb] >> PHASE[t] & 1)) {
        phm[corb] |= (unsigned char)(1u << PHASE[t]);
        hexu[HEXID[t]] = 1; ++hexcount; --deficit;
        PTRAIL[ports] = t;
        prec(t, corb, ports + 1, tok, firstlen, curlen + 1, nblk);
        ++deficit; --hexcount; hexu[HEXID[t]] = 0;
        phm[corb] &= (unsigned char)~(1u << PHASE[t]);
    }
    for (int i = 0; i < 5 && !FOUND && !CAPPED; ++i) {
        t = PAID_T[cur][i];
        if (hexu[HEXID[t]]) continue;
        int q = ORBID[t];
        int fresh = (phm[q] == 0);
        if (!fresh) { if (phm[q] >> PHASE[t] & 1) continue; if (tok < 1) continue; }
        unsigned char old = phm[q];
        phm[q] = (unsigned char)(old | (1u << PHASE[t]));
        if (fresh) { opened[nopened++] = q; deficit += 4; } else deficit -= 1;
        hexu[HEXID[t]] = 1; ++hexcount;
        PTRAIL[ports] = t;
        prec(t, q, ports + 1, tok - (fresh ? 0 : 1),
             (nblk > 1) ? firstlen : curlen, 1, nblk + 1);
        --hexcount; hexu[HEXID[t]] = 0;
        if (fresh) { --nopened; deficit -= 4; } else deficit += 1;
        phm[q] = old;
    }
}

static const char *psearch(int b, int d, int fp, int lp, int target) {
    PB = b; PDMAX = d; CD = d; PFP = fp; PLP = lp; PTARGET = target;
    NODES = 0; CAPPED = 0; FOUND = 0; PWITN = 0;
    memset(phm, 0, sizeof phm); memset(hexu, 0, sizeof hexu);
    nopened = 0; hexcount = 1; hexu[HEXID[0]] = 1;
    phm[ORBID[0]] = (unsigned char)(1u << PHASE[0]);
    opened[nopened++] = ORBID[0]; deficit = 4;
    PTRAIL[0] = 0;
    prec(0, ORBID[0], 1, b, 0, 1, 1);
    if (FOUND) return "FOUND";
    if (CAPPED) return "CAP";
    return "EXHAUSTED";
}

/* --------------------------------------------------------------- the replay */
static const char *preplay(int b, int d, int fp, int lp, int cap,
                           const int *w, int n) {
    static unsigned char ph[NQ], hx[NH];
    memset(ph, 0, sizeof ph); memset(hx, 0, sizeof hx);
    if (n != cap) return "witness length != cap";
    if (n < 1 || w[0] != 0) return "the piece must start at 123456";
    ph[ORBID[0]] = (unsigned char)(1u << PHASE[0]);
    hx[HEXID[0]] = 1;
    int norb = 1, def = 4, tok = 0, corb = ORBID[0];
    int firstlen = 0, curlen = 1, nblk = 1;
    for (int i = 0; i + 1 < n; ++i) {
        int u = w[i], t = w[i + 1];
        if (t < 0 || t >= NP) return "port out of range";
        if (hx[HEXID[t]]) return "a piece may not revisit a hexagon";
        int q = ORBID[t];
        if (ph[q] >> PHASE[t] & 1) return "phase reused";
        if (t == FREE_T[u]) {
            if (q != corb) return "clean E left its orbit";
            ph[q] |= (unsigned char)(1u << PHASE[t]);
            def -= 1; ++curlen;
        } else {
            int ok = 0;
            for (int j = 0; j < 5; ++j) if (PAID_T[u][j] == t) ok = 1;
            if (!ok) return "step is neither a clean E nor a paid joint";
            int fresh = (ph[q] == 0);
            if (!fresh) ++tok;
            ph[q] |= (unsigned char)(1u << PHASE[t]);
            if (fresh) { ++norb; def += 4; } else def -= 1;
            firstlen = (nblk > 1) ? firstlen : curlen;
            curlen = 1; ++nblk;
            corb = q;
        }
        hx[HEXID[t]] = 1;
    }
    if (tok > b) return "token budget exceeded";
    if (def > d) return "deficit budget exceeded";
    if (def != 5 * norb - n) return "deficit identity violated";
    int fl = (nblk > 1) ? firstlen : curlen;
    if (fp && fl >= 5) return "the mask requires a partial FIRST block";
    if (lp && curlen >= 5) return "the mask requires a partial LAST block";
    return NULL;
}

/* ---------------------------------------------------------------- the cells */
#define MAXP 4096
typedef struct { int b, d, fp, lp, cap, n, *w; } PCell;
static PCell PC[MAXP];
static int NPC;

static void read_pcert(const char *path) {
    FILE *f = fopen(path, "r");
    if (!f) { fprintf(stderr, "cannot open %s\n", path); exit(3); }
    char tok[64];
    int have = 0;
    while (fscanf(f, "%63s", tok) == 1) {
        if (tok[0] == '#') { int c; while ((c = fgetc(f)) != EOF && c != '\n') ; continue; }
        if (!strncmp(tok, "L6-PIECECERT", 12)) { have = 1; continue; }
        if (strcmp(tok, "pcell")) { fprintf(stderr, "expected 'pcell', saw %s\n", tok); exit(3); }
        if (NPC >= MAXP) { fprintf(stderr, "too many cells\n"); exit(3); }
        PCell *c = &PC[NPC++];
        if (fscanf(f, "%d %d %d %d %d %d", &c->b, &c->d, &c->fp, &c->lp,
                   &c->cap, &c->n) != 6) { fprintf(stderr, "bad pcell\n"); exit(3); }
        c->w = malloc(sizeof(int) * (size_t)(c->n > 0 ? c->n : 1));
        for (int i = 0; i < c->n; ++i)
            if (fscanf(f, "%d", &c->w[i]) != 1) { fprintf(stderr, "witness truncated\n"); exit(3); }
    }
    fclose(f);
    if (!have) { fprintf(stderr, "missing magic\n"); exit(3); }
}

/* Second-pass pruning, exactly as in r152/src/checker152.c: a cell that could
 * not be exhausted with the bounds available in the ascending pass is retried
 * with the bounds that pass PROVED, the cell itself excluded from its own
 * table.  The list is plain text, one "<b> <d> <cap00>" per line. */
static int EXT_B[MAXP], EXT_D[MAXP], EXT_V[MAXP], NEXT_;

static void read_pprune(const char *path) {
    FILE *f = fopen(path, "r");
    if (!f) { fprintf(stderr, "cannot open %s\n", path); exit(3); }
    while (NEXT_ < MAXP &&
           fscanf(f, "%d %d %d", &EXT_B[NEXT_], &EXT_D[NEXT_], &EXT_V[NEXT_]) == 3)
        ++NEXT_;
    fclose(f);
    fprintf(stderr, "prune list: %d (b,d) cells from %s\n", NEXT_, path);
}

static void pub_rebuild_excluding(int sb, int sd) {
    pub_init();
    for (int i = 0; i < NEXT_; ++i) {
        if (EXT_B[i] == sb && EXT_D[i] == sd) continue;
        pub_add(EXT_B[i], EXT_D[i], EXT_V[i]);
    }
}

static int do_check(const char *path, const char *prune) {
    read_pcert(path);
    if (prune) read_pprune(prune);
    uint64_t total = 0;
    int ok = 1, ncert = 0;
    printf("{\"checker\":\"P152\",\"cells\":%d,\"rows\":[\n", NPC);
    for (int i = 0; i < NPC; ++i) {
        PCell *c = &PC[i];
        if (prune) pub_rebuild_excluding(c->b, c->d);
        clock_t t0 = clock();
        const char *status, *detail = "";
        uint64_t nodes = 0;
        if (c->cap < 0) {
            /* the claim is that the mask is unreachable: exhaust everything */
            const char *r = psearch(c->b, c->d, c->fp, c->lp, 1);
            nodes = NODES; total += nodes;
            if (!strcmp(r, "FOUND")) { status = "DISAGREE"; detail = "a piece with this mask exists"; }
            else if (!strcmp(r, "CAP")) status = "UNKNOWN_CAP";
            else status = "EXACT_CERTIFIED";
        } else {
            const char *bad = c->n ? preplay(c->b, c->d, c->fp, c->lp, c->cap,
                                             c->w, c->n) : NULL;
            if (bad) { status = "WITNESS_BAD"; detail = bad; }
            else {
                const char *r = psearch(c->b, c->d, c->fp, c->lp, c->cap + 1);
                nodes = NODES; total += nodes;
                if (!strcmp(r, "FOUND")) status = "DISAGREE";
                else if (!strcmp(r, "CAP")) status = "UNKNOWN_CAP";
                else status = c->n ? "EXACT_CERTIFIED" : "UPPER_CERTIFIED";
            }
        }
        if (!strcmp(status, "EXACT_CERTIFIED") || !strcmp(status, "UPPER_CERTIFIED")) {
            ++ncert;
            if (c->fp == 0 && c->lp == 0 && c->cap >= 0) pub_add(c->b, c->d, c->cap);
        } else ok = 0;
        printf("%s{\"cell\":\"%d|%d|%d%d\",\"b\":%d,\"d\":%d,\"fp\":%d,\"lp\":%d,"
               "\"cap\":%d,\"status\":\"%s\",\"detail\":\"%s\",\"nodes\":%llu,"
               "\"seconds\":%.2f}", i ? ",\n" : "", c->b, c->d, c->fp, c->lp,
               c->b, c->d, c->fp, c->lp, c->cap, status, detail,
               (unsigned long long)nodes, (double)(clock() - t0) / CLOCKS_PER_SEC);
        fflush(stdout);
    }
    printf("\n],\"certified\":%d,\"total_nodes\":%llu,\"all_certified\":%s}\n",
           ncert, (unsigned long long)total, ok ? "true" : "false");
    return ok ? 0 : 1;
}

static int do_produce(const char *claims, const char *out) {
    int b[MAXP], d[MAXP], fp[MAXP], lp[MAXP], cl[MAXP], n = 0;
    FILE *f = fopen(claims, "r");
    if (!f) { fprintf(stderr, "cannot open %s\n", claims); return 3; }
    while (fscanf(f, "%d %d %d %d %d", &b[n], &d[n], &fp[n], &lp[n], &cl[n]) == 5)
        if (++n >= MAXP) { fprintf(stderr, "too many claims\n"); return 3; }
    fclose(f);
    for (int i = 0; i < n; ++i)                      /* heuristic prune only */
        if (fp[i] == 0 && lp[i] == 0 && cl[i] >= 0) pub_add(b[i], d[i], cl[i]);
    FILE *o = fopen(out, "w");
    if (!o) { fprintf(stderr, "cannot write %s\n", out); return 3; }
    fputs("L6-PIECECERT-1\n", o);
    fputs("# pcell <b> <d> <fp> <lp> <cap> <nports>, then the ports\n", o);
    fputs("# cap = -1 claims the mask is unreachable and carries no witness\n", o);
    int missing = 0;
    for (int i = 0; i < n; ++i) {
        int np = 0;
        if (cl[i] >= 0) {
            const char *r = psearch(b[i], d[i], fp[i], lp[i], cl[i]);
            if (!strcmp(r, "FOUND") && PWITN == cl[i]) np = PWITN;
            else { ++missing;
                fprintf(stderr, "no witness at the claim for %d|%d|%d%d "
                        "(claim %d, %s)\n", b[i], d[i], fp[i], lp[i], cl[i], r); }
        }
        fprintf(o, "pcell %d %d %d %d %d %d\n", b[i], d[i], fp[i], lp[i], cl[i], np);
        for (int j = 0; j < np; ++j) fprintf(o, "%s%d", j ? " " : "", PWIT[j]);
        fputc('\n', o);
        fflush(o);
        fprintf(stderr, "%4d/%4d %d|%d|%d%d cap=%d ports=%d\n", i + 1, n,
                b[i], d[i], fp[i], lp[i], cl[i], np);
    }
    fclose(o);
    fprintf(stderr, "cells=%d witnessed=%d upper_or_infeasible=%d\n",
            n, n - missing, missing);
    return 0;
}

int main(int argc, char **argv) {
    if (argc < 3) {
        fprintf(stderr, "usage: %s check <cert.txt> [node_cap] [prune.txt]\n"
                        "       %s produce <claims.txt> <out.txt> [node_cap]\n",
                argv[0], argv[0]);
        return 3;
    }
    common_setup();
    pub_init();
    if (!strcmp(argv[1], "check")) {
        NODECAP = argc > 3 ? strtoull(argv[3], NULL, 10) : 0;
        return do_check(argv[2], argc > 4 ? argv[4] : NULL);
    }
    if (!strcmp(argv[1], "produce")) {
        if (argc < 4) { fprintf(stderr, "produce needs an output path\n"); return 3; }
        NODECAP = argc > 4 ? strtoull(argv[4], NULL, 10) : 0;
        return do_produce(argv[2], argv[3]);
    }
    fprintf(stderr, "unknown mode %s\n", argv[1]);
    return 3;
}
